src/plugins/spaces_oss/public/plugin.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreSetup, CoreStart, PluginInitializerContext, Plugin } from 'src/core/public';
+import { SpacesOssPluginSetup, SpacesOssPluginStart } from './types';
+import { SpacesApi } from './api';
+
+export class SpacesOssPlugin implements Plugin<SpacesOssPluginSetup, SpacesOssPluginStart, {}, {}> {
+  private apiRegistered = false;
+  private api?: SpacesApi;
+
+  constructor(context: PluginInitializerContext) {}
+
+  public setup({}: CoreSetup) {
+    return {
+      registerSpacesApi: (provider: Promise<SpacesApi>) => {
+        if (this.apiRegistered) {
+          throw new Error('Spaces API can only be registered once');
+        }
+        this.apiRegistered = true;
+
+        provider.then(
+          (api) => {
+            this.api = api;
+          },
+          (error) => {
+            // eslint-disable-next-line no-console
+            console.log(
+              'Error during Spaces API promise resolution. Spaces API has been disabled',
+              error
+            );
+            this.apiRegistered = false;
+          }
+        );
+      },
+    };
+  }
+
+  public start({}: CoreStart) {
+    return {
+      isSpacesAvailable: () => this.apiRegistered,
+      getSpacesApi: () => this.api,
+    };
+  }
```

Comment:
If we remove the promise from the setup phase, then we can provide a default api implementation here, and simplify consumption for consumers a bit:

```suggestion
public start({}: CoreStart) {
  if (!this.api) {
    const error = new Error('The spaces feature is currently unavailable.');
    this.api = {
      activeSpace$: Rx.throwError(error),
      getActiveSpace: () => {
        throw error;
      },
    };
  }
  return {
    isSpacesAvailable: () => this.apiRegistered,
    getActiveSpace: this.api.getActiveSpace,
    activeSpace$: this.api.activeSpace$,
  };
}
```

Reply:
I'm hesitant to do this because we won't be able to provide the same simplified consumption / convenience for consumers when providing UI components in the future. IMO it's better to force consumers to check if Spaces is available when they are using this plugin.

Reply:
> IMO it's better to force consumers to check if Spaces is available when they are using this plugin.

They'd have to check is Spaces is available regardless of which approach we use, but I take your point.

What do you think about making the `SpacesOssPluginStart` a little more intelligent, so that we can make the ergonomics a bit easier? Something like:

```ts
export interface SpacesAvailableStartContract {
  isSpacesAvailable(): true;

  getActiveSpace(): Promise<Space>;

  activeSpace$: Observable<Space>;

  // or alternatively
  getApi(): SpacesApi;
}

export interface SpacesUnavailableStartContract {
  isSpacesAvailable(): false;

  getApi(): never;
}

export type SpacesOssPluginStart = SpacesAvailableStartContract | SpacesUnavailableStartContract;
```

If that becomes too unwieldy, then don't try to force it. Just an idea that might make it easier to consume without having to fight TypeScript

Reply:
Love it, did something very similar in dd930161.

It didn't quite work with function properties but it did work with plain boolean properties. I tried using a readonly property (with a getter) but that didn't seem to make any difference in functionality.

The only downside I can see with this approach is that we can't include TSdocs for properties while using conditional types. That seems OK in this situation though.